### PR TITLE
Solace message broker rebranding to PubSub+ updates and enhancements

### PR DESCRIPTION
**Pull request by bczoma**
- Rebranding including documentation
- Added support of using Docker hub Solace message router image
- Enhancements, fixes

### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: solace-message-router/azuredeploy.json [hunk 1]
```diff
@@ -86,38 +157,71 @@
     "imagePublisher": "OpenLogic",
     "imageOffer": "Centos",
     "OSDiskName": "osdiskfordockersimple",
+    "dataDiskName": "datadiskfordockersimple",
     "nicName": "myVMNicD",
-    "scriptUrl": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/solace-message-router/",
-    "sharedTemplatename": "shared-resources.json",
-    "solaceInstallScriptName": "deploy_vmr.sh",
+    "baseUrl": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/solace-message-router/",
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.json]: Let's update this for the BPs for the repo - ee [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: solace-message-router/azuredeploy.json [hunk 2]
```diff
@@ -86,38 +157,71 @@
...
     "deploymentModelSingleNode": {
       "numberOfNodes": 1
     },
-    "numberOfInstances": "[variables(concat('deploymentModel', parameters('deploymentModel'))).numberOfNodes]"
+    "numberOfInstances": "[variables(concat('deploymentModel', parameters('deploymentModel'))).numberOfNodes]",
+    "monitorNodeIndex": 2,
+    "lbName" : "myLB",
+    "lbID": "[resourceId('Microsoft.Network/loadBalancers',variables('lbName'))]",
+    "lbPoolName": "solace-ha-group",
+    "lbPoolID": "[concat(variables('lbID'),'/backendAddressPools/',variables('lbPoolName'))]",
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.json]: use resourceId for the id

File: solace-message-router/azuredeploy.json [hunk 3]
```diff
@@ -130,6 +234,55 @@
         }
       }
     },
+    {
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.json]: update apiVersion

File: solace-message-router/azuredeploy.json [hunk 4]
```diff
@@ -130,6 +234,55 @@
...
+            "value": "[variables('lbName')]"
+          },
+          "lbPoolName": {
+            "value": "[variables('lbPoolName')]"
+          }
+        }
+      }
+    },
+    {
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.json]: update apiVersion

File: solace-message-router/azuredeploy.json [hunk 5]
```diff
@@ -130,6 +234,55 @@
...
+      "type": "Microsoft.Resources/deployments",
+      "name": "[variables('solaceWorkspaceName')]",
+      "properties": {
+        "mode": "Incremental",
+        "templateLink": {
+          "uri": "[if(empty(parameters('workspaceName')), concat(variables('nestedUrl'), variables('workspaceDisabledSharedTemplateName')), concat(variables('nestedUrl'), variables('workspaceEnabledSharedTemplateName')))]",
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.json]: remove this pattern and just use the condition property for enable/disable

File: solace-message-router/azuredeploy.json [hunk 6]
```diff
@@ -2,69 +2,140 @@
...
+    "dnsLabelForLbIp": {
+      "type": "string",
       "metadata": {
-        "description": "Security group defined to support VMR default message vpn ports"
+        "description": "Unique DNS Label for the Public IP used to access the Load Balancer."
       }
     },
-    "dnsLabelForPublicIP": {
+    "dnsLabelForVmIp": {
       "type": "string",
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.json]: add a defaultValue here using uniqueString

File: solace-message-router/azuredeploy.json [hunk 7]
```diff
@@ -287,17 +446,43 @@
         "settings": {}
       }
     },
+    {
+      "condition": "[not(empty(parameters('workspaceName')))]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.json]: update apiVersion

File: solace-message-router/azuredeploy.parameters.json [hunk 8]
```diff
@@ -2,20 +2,20 @@
...
     },
-    "solaceVMR_URI": {
-      "value": "http://em.solace.com/DCx500W0IBC00o01dY0V0i0"
+    "dnsLabelForLbIp": {
+      "value": "GEN-UNIQUE"
+    },
+    "dnsLabelForVmIp": {
+      "value": "GEN-UNIQUE"
     },
     "deploymentModel": {
```
Comment (bmoore-msft) [inline solace-message-router/azuredeploy.parameters.json]: remove this and use a defaultValue in the template

File: solace-message-router/nestedtemplates/loadbalancer-shared-resources.json [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+      "defaultValue": "[resourceGroup().location]",
+      "metadata": {
+        "description": "Location for all resources."
+      }
+    }
+  },
+  "variables": {
+    "lbID": "[resourceId('Microsoft.Network/loadBalancers',parameters('lbName'))]",
+    "frontEndIPConfigName": "LoadBalancerFrontEnd",
+    "frontEndIPConfigID": "[concat(variables('lbID'),'/frontendIPConfigurations/',variables('frontEndIPConfigName'))]",
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/loadbalancer-shared-resources.json]: use resourceId for the ids

File: solace-message-router/nestedtemplates/loadbalancer-shared-resources.json [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    "lbPoolID": "[concat(variables('lbID'),'/backendAddressPools/',parameters('lbPoolName'))]",
+    "lbProbeName": "solace-ha-ad-health-check",
+    "lbProbeID": "[concat(variables('lbID'),'/probes/',variables('lbProbeName'))]"
+  },
+  "resources": [
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/loadbalancer-shared-resources.json]: update apiVersion

File: solace-message-router/nestedtemplates/loadbalancer-shared-resources.json [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+      "location": "[parameters('location')]",
+      "properties": {
+        "publicIPAllocationMethod": "[parameters('publicIPAddressType')]",
+        "dnsSettings": {
+          "domainNameLabel": "[parameters('dnsLabelForLbIp')]"
+        }
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/loadbalancer-shared-resources.json]: update apiVersion

File: solace-message-router/nestedtemplates/workspace-disabled-shared-resources.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/workspace-disabled-shared-resources.json]: remove this file after updating for the condition property

File: solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+  },
+  "variables": {
+    "workspaceTier": "Free",
+    "containersMarketplaceName": "Containers",
+    "syslogFacilities": [ "kern", "user", "daemon", "auth", "syslog", "uucp", "authpriv", "ftp", "cron", "local0", "local1", "local2", "local3", "local4", "local5", "local6", "local7" ]
+  },
+  "outputs": {
+    "workspaceId": {
+      "type": "string",
+      "value": "[reference(resourceId('Microsoft.OperationalInsights/workspaces/', parameters('workspaceName')), '2015-03-20').customerId]"
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json]: update apiVersion or just remove it

File: solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    },
+    "workspaceKey": {
+      "type": "string",
+      "value": "[listKeys(resourceId('Microsoft.OperationalInsights/workspaces/', parameters('workspaceName')), '2015-03-20').primarySharedKey]"
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json]: update apiVersion or just remove it

File: solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  },
+  "resources": [
+    {
+      "apiVersion": "2015-03-20",
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json]: update apiVersion

File: solace-message-router/nestedtemplates/loadbalancer-shared-resources.json (hunk 9 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/loadbalancer-shared-resources.json]: Need a bit of help how to use resourceId for the ids.

Tried:
```
    "lbID": "[resourceId('Microsoft.Network/loadBalancers',parameters('lbName'))]",
    "frontEndIPConfigName": "LoadBalancerFrontEnd",
    "frontEndIPConfigID": "[resourceId('Microsoft.Network/loadBalancers',parameters('lbName'),'frontendIPConfigurations',variables('frontEndIPConfigName'))]",
    "lbPoolID": "[resourceId('Microsoft.Network/loadBalancers',parameters('lbName'),'backendAddressPools',parameters('lbPoolName'))]",
    "lbProbeName": "solace-ha-ad-health-check",
    "lbProbeID": "[resourceId('Microsoft.Network/loadBalancers',parameters('lbName'),'probes',variables('lbProbeName'))]"
```

Got:
```
"error": {
    "code": "InvalidTemplate",
    "message": "Deployment template validation failed: 'The template variable 'frontEndIPConfigID' is not valid: Unable to evaluate template language function 'resourceId': the type 'Microsoft.Network/loadBalancers' requires '1' resource name argument(s). Please see https://aka.ms/arm-template-expressions/#resourceid for usage details.. Please see https://aka.ms/arm-template-expressions for usage details.'."
  }
```

File: solace-message-router/azuredeploy.json (hunk 1 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.json]: Updated, introduced `_artifactsLocation` and `_artifactsLocationSasToken` parameters

File: solace-message-router/azuredeploy.json (hunk 2 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.json]: Refactored - may required further work, see query for similar comment to frontEndIPConfigID.

File: solace-message-router/azuredeploy.json (hunk 3 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.json]: Updated

File: solace-message-router/azuredeploy.json (hunk 4 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.json]: Updated

File: solace-message-router/azuredeploy.json (hunk 5 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.json]: Refactored

File: solace-message-router/azuredeploy.json (hunk 6 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.json]: Added

File: solace-message-router/azuredeploy.json (hunk 7 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.json]: Updated

File: solace-message-router/azuredeploy.parameters.json (hunk 8 above)
Comment (bczoma) [reply solace-message-router/azuredeploy.parameters.json]: Removed

File: solace-message-router/nestedtemplates/loadbalancer-shared-resources.json (hunk 10 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/loadbalancer-shared-resources.json]: Updated

File: solace-message-router/nestedtemplates/loadbalancer-shared-resources.json (hunk 11 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/loadbalancer-shared-resources.json]: Updated

File: solace-message-router/nestedtemplates/workspace-disabled-shared-resources.json (hunk 12 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/workspace-disabled-shared-resources.json]: Removed

File: solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json (hunk 13 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json]: Updated, not sure of the syntax of removing

File: solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json (hunk 14 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json]: Updated, not sure of the syntax of removing

File: solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json (hunk 15 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/workspace-enabled-shared-resources.json]: Updated

**Comment by bczoma** [comment]
Hi @bmoore-msft - just wondering that the status still says Changes requested. At your convenience let me know please anything outstanding. Thank you!

File: solace-message-router/nestedtemplates/update-subnet-shared-resources.json [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+      "type": "string"
+    }
+  },
+  "variables": {},
+  "resources": [
+    {
+      "apiVersion": "2018-03-01",
+      "type": "Microsoft.Network/virtualNetworks/subnets",
+      "name": "[concat(parameters('virtualNetworkName'), '/', parameters('subnetName'))]",
+      "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline solace-message-router/nestedtemplates/update-subnet-shared-resources.json]: this needs to use the location parameter...

File: solace-message-router/nestedtemplates/loadbalancer-shared-resources.json (hunk 9 above)
Comment (bmoore-msft) [reply solace-message-router/nestedtemplates/loadbalancer-shared-resources.json]: it's a multi-part resourceID, so

resourceId('Microsoft.Network/loadbalances/frontendIPConfigurations', parameters('lbName'), variables('frontEndIPConfigName'))

File: solace-message-router/nestedtemplates/update-subnet-shared-resources.json (hunk 16 above)
Comment (bczoma) [reply solace-message-router/nestedtemplates/update-subnet-shared-resources.json]: updated

**Comment by bczoma** [comment]
@bmoore-msft , thanks for the hints, updated to using resourceIDs as suggested.

**Comment by bczoma** [comment]
Hmm, got `Unauthorized: Access is denied due to invalid credentials.` error at CI. Not sure what it is, looking into it

**Comment by bczoma** [comment]
CI passed this time

**Comment by bczoma** [comment]
@bmoore-msft please advise kindly if changes acceptable

**Comment by bczoma** [comment]
Added some updates to fix an issue detected. No major change, should not affect the Pull Request.

**Comment by bczoma** [comment]
@bmoore-msft I think now all requested changes are in

**Comment by bczoma** [comment]
Thanks @bmoore-msft !

**Pull request merged**